File: SecureChat_iOS/SecureChat/Views/SCChatSummaryView.m
//
//  SCChatSummaryView.m
//  SecureChat
//
//  Created by William Woody on 3/20/16.
//  Copyright © 2016 by William Edward Woody.
//

/*	SecureChat: A secure chat system which permits secure communications 
 *  between iOS devices and a back-end server.
 *
 *	Copyright © 2016 by William Edward Woody
 *
 *	This program is free software: you can redistribute it and/or modify it 
 *	under the terms of the GNU General Public License as published by the 
 *	Free Software Foundation, either version 3 of the License, or (at your 
 *	option) any later version.
 *
 *	This program is distributed in the hope that it will be useful, but 
 *	WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 *	or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License 
 *	for more details.
 *
 *	You should have received a copy of the GNU General Public License along 
 *	with this program. If not, see <http://www.gnu.org/licenses/>
 */

#import "SCChatSummaryView.h"

@interface SCChatSummaryView ()
@property (assign) NSInteger count;
@end

@implementation SCChatSummaryView

- (id)initWithCoder:(NSCoder *)aDecoder
{
	if (nil != (self = [super initWithCoder:aDecoder])) {
		[self internalInit];
	}
	return self;
}

- (id)initWithFrame:(CGRect)frame
{
	if (nil != (self = [super initWithFrame:frame])) {
		[self internalInit];
	}
	return self;
}

- (void)internalInit
{
	self.backgroundColor = [UIColor whiteColor];
	self.contentMode = UIViewContentModeRedraw;
}

- (void)setDeviceCount:(NSInteger)devCt
{
	self.count = devCt;
	[self setNeedsDisplay];
}

// From PaintCode
- (void)drawSummaryWithFrame: (CGRect)frame textLabel: (NSString*)textLabel
{
    //// General Declarations
    CGContextRef context = UIGraphicsGetCurrentContext();

    //// Color Declarations
    UIColor* senderColor = [UIColor colorWithRed: 0.891 green: 0.891 blue: 0.891 alpha: 1];

    //// Rectangle Drawing
    UIBezierPath* rectanglePath = [UIBezierPath bezierPathWithRect: CGRectMake(CGRectGetMinX(frame), CGRectGetMinY(frame) + CGRectGetHeight(frame) - 1, CGRectGetWidth(frame), 1)];
    [senderColor setFill];
    [rectanglePath fill];


    //// Text Drawing
    CGRect textRect = CGRectMake(CGRectGetMinX(frame), CGRectGetMinY(frame), CGRectGetWidth(frame), CGRectGetHeight(frame) - 1);
    NSMutableParagraphStyle* textStyle = NSMutableParagraphStyle.defaultParagraphStyle.mutableCopy;
    textStyle.alignment = NSTextAlignmentCenter;

    NSDictionary* textFontAttributes = @{NSFontAttributeName: [UIFont systemFontOfSize: UIFont.systemFontSize], NSForegroundColorAttributeName: UIColor.grayColor, NSParagraphStyleAttributeName: textStyle};

    CGFloat textTextHeight = [textLabel boundingRectWithSize: CGSizeMake(textRect.size.width, INFINITY)  options: NSStringDrawingUsesLineFragmentOrigin attributes: textFontAttributes context: nil].size.height;
    CGContextSaveGState(context);
    CGContextClipToRect(context, textRect);
    [textLabel drawInRect: CGRectMake(CGRectGetMinX(textRect), CGRectGetMinY(textRect) + (CGRectGetHeight(textRect) - textTextHeight) / 2, CGRectGetWidth(textRect), textTextHeight) withAttributes: textFontAttributes];
    CGContextRestoreGState(context);
}

- (void)drawRect:(CGRect)rect
{
	NSString *str;
	if (self.isSelf) {
		if (self.count == 1) {
			str = NSLocalizedString(@"You have %d device.", @"single");
		} else {
			str = NSLocalizedString(@"You have %d devices.", @"single");
		}
	} else {
		if (self.count == 1) {
			str = NSLocalizedString(@"User has %d device.", @"single");
		} else {
			str = NSLocalizedString(@"User has %d devices.", @"single");
		}
	}

	str = [NSString stringWithFormat:str, (int)self.count];
	[self drawSummaryWithFrame:self.bounds textLabel:str];
}

@end

File: SecureChat_android/app/src/main/java/com/chaosinmotion/securechat/views/SCChatSummaryView.java
/*
 * Copyright (c) 2016. William Edward Woody
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>
 *
 */

package com.chaosinmotion.securechat.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

import com.chaosinmotion.securechat.R;

/**
 * Displays the number of views within this view.
 * Created by woody on 4/20/16.
 */
public class SCChatSummaryView extends View
{
	private TextPaint paint;
	private boolean isSelf;
	private int deviceCount;

	public SCChatSummaryView(Context context)
	{
		super(context);
		internalInit();
	}

	public SCChatSummaryView(Context context, AttributeSet attrs)
	{
		super(context, attrs);
		internalInit();
	}

	public SCChatSummaryView(Context context, AttributeSet attrs, int defStyleAttr)
	{
		super(context, attrs, defStyleAttr);
		internalInit();
	}

	public void setSelf(boolean f)
	{
		isSelf = f;
	}

	public void setDeviceCount(int dc)
	{
		deviceCount = dc;
	}

	private void internalInit()
	{
		float density = getResources().getDisplayMetrics().density;
		paint = new TextPaint(Paint.ANTI_ALIAS_FLAG);
		paint.setTextSize((int)(density * 13));
		paint.setColor(Color.GRAY);
		paint.setTextAlign(Paint.Align.CENTER);
	}

	@Override
	protected void onDraw(Canvas canvas)
	{
		super.onDraw(canvas);

		String format;
		String str;
		if (isSelf) {
			if (deviceCount == 1) {
				format = getResources().getString(R.string.you_have_singular);
			} else {
				format = getResources().getString(R.string.you_have_plural);
			}
		} else {
			if (deviceCount == 1) {
				format = getResources().getString(R.string.user_has_singular);
			} else {
				format = getResources().getString(R.string.user_has_plural);
			}
		}
		str = String.format(format,deviceCount);

		Paint.FontMetrics fm = paint.getFontMetrics();
		int ypos = (int)(getHeight() - fm.ascent - fm.descent)/2;
		int xpos = (int)(getWidth()/2);
		canvas.drawText(str,xpos,ypos,paint);

		canvas.drawRect(0,getHeight()-1,getWidth(),getHeight(),paint);
	}
}
